lib/sources.nix

```diff
@@ -101,4 +101,201 @@ rec {
+
+        # For fast non-O(n) lookup, we turn `srcDirs` and `srcFiles` into
+        # string-keyed attrsets first.
+        srcDirsSet = keySet (map toString srcDirs);
+        srcFilesSet = keySet (map toString srcFiles);
+
+        # We also turn `srcDirs` into a directory-prefix-tree so that we can
+        # check whether a given path is under one of the `srcDirs` in sub-O(n).
+        srcDirsTree = pathComponentsToTree (map splitPath srcDirs);
+      in
+        # The actual filter function with per-file processing
+        fullPath: type:
+          let
+            fileName = baseNameOf (toString fullPath);
+
+            components = splitPath fullPath;
+
+            isExplicitSrcFile = builtins.hasAttr fullPath srcFilesSet;
+            isExplicitSrcDir = type == "directory" && builtins.hasAttr fullPath srcDirsSet;
+            # The below is equivalent to
+            #   any (srcDir: hasPrefix (toString srcDir + "/") fullPath) srcDirs;
+            # but faster than O(n) where n is the number of `srcDirs` entries.
+            isUnderSomeSrcDir = isPrefixOfLeafPath components srcDirsTree;
+
+            isHidden = excludeHidden && hasPrefix "." fileName;
+
+            hasExcludedComponentInPath = any (c: elem c pathComponentExcludes) components;
+
+            isSourceInput =
+              isExplicitSrcFile ||
+              isExplicitSrcDir ||
+              (isUnderSomeSrcDir && !isHidden && !hasExcludedComponentInPath);
+
+            tracing =
+              let
+                prefix = if debugTracePrefix == "" then "" else debugTracePrefix + ": ";
+                action = if isSourceInput then "include" else "skip   ";
+                # Pad "regular" to be as wide as "directory" for aligned output.
+                formattedType = if type == "regular" then "regular  " else type;
+              in
+              if debugEnableTracing
+                then builtins.trace "${prefix}${action} ${formattedType} ${fullPath}"
+                else id;
+          in
+            tracing isSourceInput;
+
+  # A general-purpose, explicit source code importer suitable for most
+  # packaging needs.
+  #
+  # See `explicitSourceFilter` for an explanation of the filter arguments.
+  #
+  # Example usage:
+  #
+  #    src = lib.explicitFilterSource "mypackage" ./. {
+  #      srcDirs = [
+  #        ./src
+  #        ./app
+  #        ./images
+  #      ];
+  #      srcFiles = [
+  #        ./mypackage.cabal
+  #        ./Setup.hs
+  #      ];
+  #      pathComponentExcludes = [ "build" "gen" ];
+  #    };
+  #
+  # Note that `srcDirs = [ ./. ]` is also permitted.
+  #
+  # But consider that in most cases you will not want to include files
+  # that are not relevant for the build, such as `.nix` files, so that
+  # input hashes do not change unnecessarily.
+  #
+  # If you want to combine it with other filters, use `explicitSourceFilter`
+  # directly instead.
+  explicitFilterSource = name: topPath: filterArgs:
+    builtins.path {
```

Comment:
This will benefit from calling `cleanSourceWith` so you can compose (intersect) with other filters, such as [gitignore](https://github.com/hercules-ci/gitignore/).
I suppose you've avoided that function because it wasn't always deterministic in the `name`, but that can be fixed with #67996

Reply:
Done.